ray_maple_workflow.py

```diff
@@ -0,0 +1,130 @@
+        "read and write to the google cloud storage buckets.",
+    )
+
+    parser.add_argument(
+        "--adc_dir",
+        required=False,
+        default="",
+        help="The directory path for application default credentials (adc). This path must be set if "
+        "you want to give ray access to your gcs buckets when you are running this workflow on your "
+        "*local computer*. It is necessary for service account impersonation, which is used to give "
+        "this code access to your storage bucket when running the code locally."
+    )
+
+    parser.add_argument(
+        "--weight_file",
+        required=False,
+        default="hyp_best_train_weights_final.h5",
+        help="The file path to where the model weights can be found. Should be "
+        "relative to the root directory.",
+    )
+
+    parser.add_argument(
+        "--gpus_per_core",
+        required=False,
+        default=1,
+        help="Number of GPUs available per core. Used to determine how many "
+        "inference processes to spin up. Set this to 0 if you want to run the "
+        "workflow on a CPU.",
+        type=int
+    )
+
+    args = parser.parse_args()
+
+    image_name = args.image
+    config = MPL_Config(
+        args.root_dir, args.adc_dir, args.weight_file, num_gpus_per_core=args.gpus_per_core
+    )
+
+    print("0. Load geotiffs into ray dataset")
+    dataset = create_geotiff_images_dataset(config).map(add_image_name)
+    print("Ray dataset schema:", dataset.schema())
+    print("1. Start calculating watermask")
+    dataset_with_water_mask = dataset.map(fn=ray_image_preprocessing.cal_water_mask,
+                                          fn_kwargs={"config": config})
+    print("Dataset schema with water mask: ", dataset_with_water_mask.schema())
+    print("2. Start tiling image")
+    image_tiles_dataset = dataset_with_water_mask.flat_map(
+        fn=ray_tile_and_stitch_util.tile_image, fn_kwargs={"config": config})
+    image_tiles_dataset = image_tiles_dataset.drop_columns(["mask"])
+    print("Dataset schema with image tiles: ", image_tiles_dataset.schema())
+    print("3. Start inferencing")
+    inferenced_dataset = image_tiles_dataset.map(
+        fn=ray_inference.MaskRCNNPredictor, fn_constructor_kwargs={"config": config}, concurrency=2)
+    print("Dataset schema with inferenced tiles: ", inferenced_dataset.schema())
+    print("4. Start stitching")
+    data_per_image = inferenced_dataset.groupby(
+        "image_name").map_groups(ray_tile_and_stitch_util.stitch_shapefile)
+    print("Dataset schema where each row is an image (result of a group by tile): ",
+          data_per_image.schema())
+    print("5. Write shapefiles")
+    shapefiles_dataset = data_per_image.map(
+        fn=ray_write_shapefiles.WriteShapefiles, fn_constructor_kwargs={"config": config}, concurrency=2)
+    print("Done writing shapefiles", shapefiles_dataset.schema())
```

Comment:
In most of the map calls, concurrency is not being set. I think it is important to make a note that the concurrency parameter should be tuned based on the resources available in the raycluster in order to make the best use of the compute resources available. If no concurrency is set, it just runs a single worker for it which won't provide much improvement.

Reply:
added an arg for concurrency